src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -83,8 +85,34 @@ public Http2Connection(HttpConnectionContext context)
         // Capture the ExecutionContext before dispatching HTTP/2 middleware. Will be restored by streams when processing request
         _context.InitialExecutionContext = ExecutionContext.Capture();
 
+        var inputPipeOptions = new PipeOptions(pool: context.MemoryPool,
```

Comment:
Does this introduce a new pipe... That's unfortunate.

Reply:
It does introduce a new pipe. This part of the PR description describes how we could avoid that.

> This does introduce an extra copy similar to what we already have for HTTP/2 input, but the benchmark results clearly show this is worthwhile in order to offload the TLS work to a thread that doesn't block other HTTP/2 streams. We could avoid this copy by updating `ConcurrentPipeWriter` to dispatch calls to `FlushAsync` and `WriteAsync`. I didn't do that for this initial iteration because we'd want to use a pooled `IValueTaskSource` to support this. We'd also want to make `ConcurrentPipeWriter` aware of the `MaxResponseBufferSize` so it wouldn't *always* return incomplete ValueTasks for dispatched writes.

I wanted to keep this simple for now though so we can possibly service this. Long term we can do the custom `PipeWriter`, or better yet get rid of the `_writeLock` altogether using Channels.

Reply:
I don't feel like a lot of time should be invested in what should hopefully be replaced by the best solution (channels).